Bottom: 9c8c886a489050f572182efc391e495116753f62
Top:    507b5d31b7a10383a748e34cb859ed5d398ca84e
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-13 11:04:04 -0500

Add codebase generator function

The codebase generator takes information about a change, like repository,
branch, and revision, and creates a predictable string describing the code.
This is used to decide upon the builder for a change and also to tie a fork
repository back to its parent for merge requests.


---
diff --git a/docker/buildbot-master/buildbot/master.cfg b/docker/buildbot-master/buildbot/master.cfg
index 02056a49..2b93069b 100644
--- a/docker/buildbot-master/buildbot/master.cfg
+++ b/docker/buildbot-master/buildbot/master.cfg
@@ -139,3 +139,28 @@ c['workers'].extend(ptxdist.workers)
 c['schedulers'].extend(ptxdist.schedulers)
 c['builders'].extend(ptxdist.builders)
 c['change_source'].extend(ptxdist.change_source)
+
+
+def codebaseGenerator(chdict):
+    """ Generate a codebase description.
+    
+    The codebase generator takes information about a change, like repository,
+    branch, and revision, and creates a predictable string describing the code.
+    This is used to decide upon the builder for a change and also to tie a fork
+    repository back to its parent for merge requests.
+    """
+    pprint(chdict, indent=2)
+    path = urlparse(chdict['revlink']).path.lstrip('/')
+
+    components = []
+    while path != "":
+        path, tail = os.path.split(path)
+        if tail != "":
+            components.append(tail)
+    components.reverse()
+
+    codebase = os.path.join(*components[0:2])
+    return codebase
+
+
+c['codebaseGenerator'] = codebaseGenerator
